x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/package-info.java

```diff
@@ -0,0 +1,168 @@
+ * Used in `STATS` and similar commands.
+ *
+ * <h2>Guide to adding new aggregate function</h2>
+ * <ol>
+ *     <li>
+ *         Aggregation functions are more complex than scalar functions, so it's a good idea to discuss
+ *         the new function with the ESQL team before starting to implement it.
+ *         <p>
+ *             You may also discuss its implementation, as aggregations may require special performance considerations.
+ *         </p>
+ *     </li>
+ *     <li>
+ *         Run the csv tests (see {@code x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java})
+ *         from within Intellij or, alternatively, via Gradle:
+ *         {@code ./gradlew :x-pack:plugin:esql:test --tests "org.elasticsearch.xpack.esql.CsvTests"}
+ *         IntelliJ will take a few minutes to compile everything but the test itself should take only a few seconds.
+ *         This is a fast path to running ESQL's integration tests.
+ *     </li>
+ *     <li>
+ *         Pick one of the csv-spec files in {@code x-pack/plugin/esql/qa/testFixtures/src/main/resources/}
+ *         and add a test for the function you want to write. These files are roughly themed but there
+ *         isn't a strong guiding principle in the organization.
+ *     </li>
+ *     <li>
+ *         Rerun the {@code CsvTests} and watch your new test fail.
+ *     </li>
+ *     <li>
+ *         Find an aggregate function in this package similar to the one you are working on and copy it to build
+ *         yours.
+ *         Your function might extend from the available abstract classes. Check the javadoc of each before using them:
+ *         <ul>
+ *             <li>
+ *                 {@link org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction}: The base class for aggregates
+ *             </li>
+ *             <li>
+ *                 {@link org.elasticsearch.xpack.esql.expression.function.aggregate.NumericAggregate}: Aggregation for numeric values
+ *             </li>
+ *             <li>
+ *                 {@link org.elasticsearch.xpack.esql.expression.function.aggregate.SpatialAggregateFunction}:
+ *                 Aggregation for spatial values
+ *             </li>
+ *         </ul>
+ *     </li>
+ *     <li>
+ *         Fill the required methods in your new function. Check their JavaDoc for more information.
+ *         Here are some of the important ones:
+ *         <ul>
+ *             <li>Constructor: Review the constructor annotations, and make sure to add the correct types and descriptions</li>
+ *             <li>
+ *                 {@code resolveType}: Check the metadata of your function parameters.
+ *                 This may include types, if they are foldable, or their possible values.
+ *             </li>
+ *             <li>
+ *                 {@code dataType}: This will return the datatype of your function.
+ *                 May be based on its current parameters.
```

Comment:
You are allowed to call `resolveType` in this method. It's not read-only - just idempotent.

Reply:
I thought resolveType was always called before dataType (?). Which makes calling it again a no-op (Specially since it doesn't return anything interesting at this point)

Reply:
I think it is always is called before `dataType`, but we tend to be a bit permissive about this one and call it again. This code isn't on the hot path and it makes writing tests a bit easier. At least, I think that's what's up.